irksome/nprk.py

```diff
@@ -0,0 +1,328 @@
+"""Classes to implement non-linearly partitioned Runge-Kutta (NPRK) schemes based on
+Buvoli and Southworth."""
+import numpy as np
+
+from firedrake import Function, split, \
+    NonlinearVariationalProblem, NonlinearVariationalSolver
+
+from ufl.form import Form
+from ufl.algorithms.analysis import extract_type
+
+from .deriv import TimeDerivative
+from .tools import MeshConstant, replace
+from .manipulation import extract_terms
+from .nprk_tableaux import NPRKTableau
+
+
+def _has_len(x) -> bool:
+    """Checks if an object has a length"""
+
+    try:
+        _ = len(x)
+    except Exception:
+        return False
+
+    return True
+
+
+def _check_aux_t_dep(F: Form, u0: Function, u1: Function, u2: Function) -> tuple[np.ndarray]:
+    """Identifies which subfunctions are auxiliary (no time derivative)
+    and which are time-dependent. Also checks that variables are exactly
+    one of the two and that auxiliary variable do not appear in a partitioned
+    argument."""
+
+    num_fields = len(u0.function_space())
+
+    if num_fields == 1:
```

Comment:
If you check the updated manipulation that landed not too long ago, you'll see that we've moved away from this approach to iterating over `u0` itself: `u0bits = u0[i] for i in numpy.ndindex(u.ufl_shape)`.  This is a different path that doesn't have so many branches on mixed vs vector vs whatever.

Reply:
We don't even have to worry about mixed vs vector these days. We just preprocess the form with `expand_time_derivatives`.